### PR TITLE
Import elections from elections.democracyclub.org.uk

### PR DESCRIPTION
**Pull request by symroe**

### CODE REVIEW

**Comment by symroe** [comment]
[EE -> YNR import script](https://trello.com/c/RK1AGJj9/882-ee-ynr-import-script)

File: elections/uk/management/commands/uk_create_elections_from_every_election.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            area_name = area_info['division']['name']
+            post_role = area_info['division']['geography_curie']
+            post_extra_slug = ":".join([
+                area_info['division']['division_type'],
+                area_info['division']['geography_curie'].split(':')[-1]
+            ])
+            area_identifier = area_info['division']['geography_curie']
+            winner_count = area_info['division']['seats_total']
+        else:
+            # TODO
```
Comment (mhl) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: What's still to do here?

File: elections/uk/management/commands/uk_create_elections_from_every_election.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            organization_extra = OrganizationExtra.objects.create(
+                base=organization,
+                slug=org_slug
+            )
+        if not election_model.organization:
+            election_model.organization = organization_extra.base
+            election_model.save()
+        return organization_extra
+
+    def add_area(self, area_info, election, organization_extra):
```
Comment (mhl) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: I found this method quite hard to follow compared to the others, at least partly because it's quite long. Some of it could be factored out into methods (e.g. `get_party_set(territory_code)` for the first lines, or `get_area_type` for the next).

The bits that are based on `area_info['division']` could maybe do with a comment or two about what it's doing? Since I'm not very familiar with what these values from the API mean, that would help me anyway. (I couldn't find documentation of them easily on https://elections.democracyclub.org.uk but if there's somewhere that could be linked to, that might help too.)

File: elections/uk/management/commands/uk_create_elections_from_every_election.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            country = "ni"
+        else:
+            partyset_name = "Great Britain"
+            country = "gb"
+
+        party_set, _ = PartySet.objects.get_or_create(
+            slug=country, defaults={'name': partyset_name}
+        )
+
+        if area_info['division']:
```
Comment (mhl) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: I think that currently there's a strong assumption for the UK that the `name` of an `AreaType` corresponds to a MapIt `Type` - I just wanted to check that that's still the case (or it might be worth a comment in the code to that effect).

File: elections/uk/management/commands/uk_create_elections_from_every_election.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                # 'parent': parent_area,
+            }
+        )
+
+        AreaExtra.objects.get_or_create(
+            base=area,
+            defaults={'type': area_type}
+        )
+
+        post, _ = Post.objects.update_or_create(
```
Comment (mhl) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: I think the creation or update of the post has to be done a bit differently, I'm afraid, since effectively the primary key for a Post + PostExtra pair is the `slug` field on `PostExtra`. The code here risks unpairing those (I think!). The pattern for doing this safely (which should probably be extracted to a manager method to make this easier, e.g. [something like what we did with `ImageExtraManager`](https://github.com/DemocracyClub/yournextrepresentative/blob/c668ede5f5c25f93228238700403d071f04d6f4f/candidates/models/popolo_extra.py#L745-L780)  does) is:

https://github.com/DemocracyClub/yournextrepresentative/blob/c668ede5f5c25f93228238700403d071f04d6f4f/elections/cr/management/commands/cr_create_basic_site.py#L115-L138

File: elections/uk/management/commands/uk_create_elections_from_every_election.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            defaults={
+                'role': post_role,
+                'label': area_name,
+            })
+
+        post_extra, _ = PostExtra.objects.update_or_create(
+            slug=post_extra_slug,
+            defaults={
+                'base': post,
+                'party_set': party_set,
```
Comment (mhl) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: I think the `group` should be set here as well - at the moment it should be one of `Scotland`, `England`, `Wales` or `Northern Ireland`.

File: elections/uk/management/commands/uk_create_elections_from_every_election.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+        return Election.objects.update_or_create(
+            slug=election_id,
+            election_date=election_date,
+            current=current,
+            defaults={
+                "name": election_dict['election_title'],
+                "for_post_role": election_dict['elected_role'],
+                "candidate_membership_role": "Candidate",
+                "show_official_documents": True,
```
Comment (symroe) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: This is a good point, and needs a new feature in EE https://trello.com/c/JGrBsXj9/895-support-voting-systems-in-ee

File: elections/uk/management/commands/uk_create_elections_from_every_election.py (hunk 3 above)
Comment (symroe) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: This will be the case sometimes, as elections.DC will use mapit.mysociety.org code where possible, but mySociety's mapit doesn't support all the area types that have elections, so we have to use our own. Related, I think I will also move away from using mapit for looking up area types (in a near future PR)

Comment (symroe) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: Election areas that mapit.mysociety.org doesn't support: PCC, (some) combined authorities, any area with boundaries that haven't been added to BoundaryLine before the elections.

File: elections/uk/management/commands/uk_create_elections_from_every_election.py (hunk 1 above)
Comment (symroe) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: Comment in the wrong place, moved and a note added.

File: elections/uk/management/commands/uk_create_elections_from_every_election.py (hunk 3 above)
Comment (symroe) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: A bit more context here: https://github.com/DemocracyClub/EveryElection/issues/1

File: elections/uk/management/commands/uk_create_elections_from_every_election.py (hunk 6 above)
Comment (symroe) [inline elections/uk/management/commands/uk_create_elections_from_every_election.py]: Added in 99b21bb

**Comment by mhl** [comment]
Hi @symroe - just a couple of other suggestions: since this is doing some difficult-to-get-right operations on the paired models, I wondered whether you might consider reviewing + merging this first: https://github.com/DemocracyClub/yournextrepresentative/pull/43 and then adding something like:

```python
        errors = check_paired_models()
        if errors:
            for error in errors:
                print(error)
            raise Exception("Unpaired models found - raising an exception to rollback the transaction")
```

... at the end of the `handle` method?

One other thing I thought I'd left a comment about, but I can't find now :confused: is the possibility to change the end condition of `while` loop in the `handle` method so that it reduces to:

```python
        url = "{}api/elections?group_type=organisation".format(
            self.EE_BASE_URL)
        while url:
            req = requests.get(url)
            data = req.json()
            self.process_results(data['results'])
            url = data.get('next')
```

**Comment by mhl** [comment]
:+1:
